bootstrap/api/v1beta1/rke2config_webhook.go

```diff
@@ -37,20 +38,43 @@ var (
 	rke2configlog         = logf.Log.WithName("rke2config-resource")
 )
 
-// SetupWebhookWithManager sets up and registers the webhook with the manager.
-func (r *RKE2Config) SetupWebhookWithManager(mgr ctrl.Manager) error {
+// RKE2ConfigCustomDefaulter struct is responsible for setting default values on the custom resource of the
+// Kind RKE2Config when those are created or updated.
+// NOTE: The +kubebuilder:object:generate=false marker prevents controller-gen from generating DeepCopy methods,
+// as it is used only for temporary operations and does not need to be deeply copied.
+// +kubebuilder:object:generate=false
+type RKE2ConfigCustomDefaulter struct{}
+
+// RKE2ConfigCustomValidator struct is responsible for validating the RKE2Config resource
+// when it is created, updated, or deleted.
+// NOTE: The +kubebuilder:object:generate=false marker prevents controller-gen from generating DeepCopy methods,
+// as it is used only for temporary operations and does not need to be deeply copied.
+// +kubebuilder:object:generate=false
+type RKE2ConfigCustomValidator struct{}
+
+// SetupRKE2ConfigWebhookWithManager sets up the Controller Manager for the Webhook for the RKE2ControlPlaneTemplate resource.
+func SetupRKE2ConfigWebhookWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewWebhookManagedBy(mgr).
-		For(r).
+		For(&RKE2Config{}).
+		WithValidator(&RKE2ConfigCustomValidator{}).
+		WithDefaulter(&RKE2ConfigCustomDefaulter{}).
 		Complete()
 }
 
 //+kubebuilder:webhook:path=/mutate-bootstrap-cluster-x-k8s-io-v1beta1-rke2config,mutating=true,failurePolicy=fail,sideEffects=None,groups=bootstrap.cluster.x-k8s.io,resources=rke2configs,verbs=create;update,versions=v1beta1,name=mrke2config.kb.io,admissionReviewVersions=v1
 
-var _ webhook.Defaulter = &RKE2Config{}
+var _ webhook.CustomDefaulter = &RKE2ConfigCustomDefaulter{}
 
 // Default implements webhook.Defaulter so a webhook will be registered for the type.
-func (r *RKE2Config) Default() {
-	DefaultRKE2ConfigSpec(&r.Spec)
+func (r *RKE2ConfigCustomDefaulter) Default(_ context.Context, obj runtime.Object) error {
+	rc, ok := obj.(*RKE2Config)
+	if !ok {
+		return fmt.Errorf("expected a RKE2Config object but got %T", obj)
```

Comment:
how about:

```suggestion
		return apierrors.NewBadRequest(fmt.Sprintf("expected a RKE2Config but got a %T", obj))
```

since we are dealing with webhook and it can properly handle a `400 Bad Request` response in more structured way? (in all other occurences)